[클린코드 2기 정현수] 자판기 미션 STEP1

junghyeonsu opened the pull request:
안녕하세요 리뷰어님! STEP1 상품 관리 탭에 대한 기능들 구현 완료해서 리뷰 요청드립니다.

제가 맨 처음에 `MVC` 패턴을 이용해서 구현을 하려고 구조를 잡고 있었는데,
처음부터 너무 구조에 신경쓰고, 좀 딱딱하게 구조를 나눠놔서 그런지
여러 상태들을 어떻게 전달을 하고, 관리를 해야하는지 감이 잘 잡히지 않아서
다시 싹 다 초기화 시키고, 다시 코드를 짜서 현재의 구조가 됐습니다.

지금 궁금한 것은 다음과 같습니다.
nextstep이 거의 끝나가는데, 제가 면접도 보고(핑계) 이것저것 하느라 열의적으로 참여를 하지 못한 것도 있지만
nextstep을 진행하면서 여러 미션을 진행을 했는데, 뭔가 내가 잘해졌다? 라는 것 보다 어떤 식으로 짜야할지
잘 모르겠는,, 그런 혼란스러움이 가득합니다.

현재 `VendingMachine`에서 선언된 `state`에 따라서 `state`가 변하면 `setState`를 해주어서 다시 렌더링을 하는 방식으로 진행을 했는데, 이게 앱이 더욱 커지면 `VendingMachine`이 더 커질 것 같아서 도메인마다 모델들을 분리를 시키고 싶은데
만약 분리를 한다면, `setState` 부분을 어떻게 처리를 해주어야 하는지 감이 안잡힙니다.

그러니까 `currentTab`을 관리하는 `model`과 `products`를 관리하는 `model`로 따로 관리를 하고 싶은데,
현재로서는 어떤식으로 나누어야 하는지 잘 감이 안옵니다.

혹시 도움을 얻을 수 있을까요?..

### [데모 링크](https://junghyeonsu.github.io/js-vending-machine/)

## ❗ 요구사항

### 상품 관리 탭

- [x] 각 탭을 클릭했을 때 알맞은 View를 보여준다.
- [x] 현재 탭을 `localStorage`에 저장해놓고, 새롭게 나갔다 들어와도 유지된다.
- [x] 처음 들어갔을 때 상품명, 금액, 수량을 입력할 수 있는 input이 존재한다.
- [x] 상품명, 금액, 수량은 공백이 불가능하다.
- [x] 수량은 `1`이상인 정수여야 한다.
- [x] 가격은 `100`이상인 정수여야 한다.
- [x] 가격은 `10`으로 나누어 떨어져야 한다.
- [x] 같은 상품명으로 추가시 기존의 데이터를 덮어씌운다.
- [x] 추가된 상품은 이름, 가격, 수량 순으로 테이블에 추가된다.
- [x] 추가된 상품은 `localStorage`를 이용해서 데이터를 유지시킨다.

junghyeonsu committed:
chore: package.json 파일 생성

junghyeonsu committed:
chore: prettier, eslint 설치

junghyeonsu committed:
chore: cypress 설정

- cypress
- cypress-watch-and-reload

junghyeonsu committed:
chore: cypress scripts 추가

junghyeonsu committed:
chore: index.js 생성

junghyeonsu committed:
chore: cypress.json 수정

- watch 파일 경로 수정

junghyeonsu committed:
docs: step1 요구사항 리스트 작성

junghyeonsu committed:
feat: 자판기 처음들어갔을 때 처리

- 상품 추가하기 화면 뜨도록

junghyeonsu committed:
chore: ProductManagerView selector 상수화 처리

junghyeonsu committed:
docs: 요구사항 추가 및 해결

junghyeonsu committed:
chore: ProductManager 이름 변경

- ProductManage로 변경

junghyeonsu committed:
test: 메뉴 선택 테스트 생성

junghyeonsu committed:
feat: Tab 컨트롤러 & 모델 생성

- 현재의 탭을 관리하는 컨트롤러 생성
- 현재의 탭에 대한 데이터를 관리하는 모델 생성

junghyeonsu committed:
feat: 각 탭 클릭 시 이동 구현

- 각 탭 클릭 시 알맞은 view 보여주도록 구현
- 현재 탭은 localStorage에 저장됨
- 처음 로딩시 localStorage에 값이 있으면 들고오고 아니면 '상품관리'탭이 기본값으로 지정

junghyeonsu committed:
docs: 완료 기능 체크

junghyeonsu committed:
chore: 코드 간략화

junghyeonsu committed:
chore: cypress 설정 수정

- cypress-watch-and-reload가 바라보는 폴더 추가

junghyeonsu committed:
test: 상품 추가 탭에 대한 테스트 작성

junghyeonsu committed:
feat: 상품명에 대한 검증 구현

- 상품명이 비어있을 때 경고창 띄움

junghyeonsu committed:
feat 상품 가격에 대한 검증 구현

- 상품 가격은 필수값
- 상품 가격은 100이상 정수
- 상품 가격은 10으로 나누어 떨어지는 정수

junghyeonsu committed:
feat: 상품 수량에 대한 검증 구현

- 상품 수량은 필수값
- 상품 수량은 1이상인 정수

junghyeonsu committed:
docs: 구현 완료 체크

junghyeonsu committed:
test: 상품 추가 테스트 구현

junghyeonsu committed:
feat: localStorage 관리 객체 수정

junghyeonsu committed:
refactor: 파일 및 폴더 구조 변경

- 단순히 vendingMachine과 관련된 인스턴스로 구현

junghyeonsu committed:
test: 상품 중복 추가에 대한 테스트 구현

junghyeonsu committed:
feat: 상품 중복 추가시 처리

- 중복되어있던 데이터를 삭제 후, 다시 추가

junghyeonsu committed:
refactor: validator 생성

junghyeonsu committed:
refactor: Validator 인스턴스로 수정

- "검증을 해주는 객체" 라는 의미로 인스턴스 생성

junghyeonsu committed:
refactor: VendingMachine 코드 정리

- addEvents와 initEvents를 initEvents로 하나로 합침
- onClickMenu를 changeTab으로 변경

junghyeonsu committed:
chore: VendingMachineView render 메소드명 변경

- render에서 renderTab으로 조금 더 직관적으로 변경

junghyeonsu committed:
docs: 완료 기능 체크

junghyeonsu committed:
refactor: else if문을 if문으로 수정

- else if는 넣을 필요 없다고 생각해서 수정

EungyuCho commented on src/VendingMachine.js:
```diff
@@ -0,0 +1,83 @@
+        inputProductName,
+        inputProductPrice,
+        inputProductQuantity,
+      );
```
```suggestion
    const product = {
      name: $(`#${SELECTOR.PRODUCT_NAME_INPUT_ID}`).value,
      price: $(`#${SELECTOR.PRODUCT_PRICE_INPUT_ID}`).value,
      quantity: $(`#${SELECTOR.PRODUCT_QUANTITY_INPUT_ID}`).value
    }

    try {
      this.validator.validateProductDatas(product);
```
인자를 [객체](https://github.com/qkraudghgh/clean-code-javascript-ko#%ED%95%A8%EC%88%98functions)형태로 받으면 더 좋을거같아요.
아래에 있는 setState에서도 그대로 사용할 수 있어서 좋아보여요.

EungyuCho commented on src/constants.js:
```diff
@@ -0,0 +1,31 @@
+  REQUIRED_PRODUCT_QUANTITY: '상품 수량을 입력해주세요.',
+  PRODUCT_PRICE_HAVE_TO_OVER_100: '상품 금액은 100이상인 정수여야 합니다.',
+  PRODUCT_PRICE_HAVE_TO_DIVIDED_BY_10: '상품 금액은 10으로 나누어 떨어져야 합니다.',
+  PRODUCT_QUANTITY_HAVE_TO_OVER_1: '상품 수량은 1이상인 정수여야 합니다.',
```
정책에 따라 바뀔 수 있는 여지가 있는 숫자에 대해서는 상수로 분리해도 좋을거같아요.
만약 100원에서 1000원으로 바뀐다면 코드라인에서 100으로 작성되어있는 가격 정보를 일일이 바꿔줘야하고 해당 상수의 변수명과 텍스트까지 직접 바꿔야돼서 힘들 것 같아요 💦 
혹시라도 절대 정책이 안변한다고 확신하고 코드를 작성하신거라면 넘어가셔도 무방합니다 🙂

EungyuCho commented on src/VendingMachineView.js:
```diff
@@ -0,0 +1,23 @@
+
+  renderTab(state) {
+    this.target.innerHTML = this.changeTab[state.currentTab](state);
+  }
```
history API를 사용해보시는건 어떨까요?
내부 콘텐츠가 다르기 때문에 새로운 탭을 클릭했을 때 사용자는 다른 페이지를 열었다고 인식할 것 같아요.
상품관리, 잔돈 충전, 구매 페이지 등 서로 내부 콘텐츠가 다르기 때문에 라우팅을 처리해주는게 바람직하다고 생각해요.

여유가 되신다면 hashChange를 이용한 방법까지 확인해보시고 history와 어떤 차이가 있는지 알아보시면 좋을거같아요.

EungyuCho commented on src/templates/ProductManageTemplate.js:
```diff
@@ -0,0 +1,40 @@
+    <tbody id="${SELECTOR.PRODUCT_INVENTORY_CONTAINER_ID}">
+      ${state.products
+        .map(product => productTemplate(product.name, product.price, product.quantity))
+        .join('')}
```
```suggestion
        .map(productTemplate)
        .join('')}
```
`productTemplate`도 인자들을 객체 형태로 받으면 이렇게 바꿀 수 있겠네요!

EungyuCho commented on src/templates/index.js:
```diff
@@ -0,0 +1,5 @@
+import ProductPurchaseTemplate from './ProductPurchaseTemplate.js';
+import VendingMachineManageTemplate from './VendingMachineManageTemplate.js';
+
+export { ProductManageTemplate, ProductPurchaseTemplate, VendingMachineManageTemplate };
```
import Path 관리도 신경써주셨네요 👍 👍

EungyuCho commented on src/VendingMachine.js:
```diff
@@ -0,0 +1,83 @@
+    this.state = {
+      currentTab: store.getValue(STORE_KEY.CURRENT_TAB) || TAB.PRODUCT_MANAGE_TAB,
+      products: store.getValue(STORE_KEY.PRODUCTS) || [],
+    };
```
본문 코멘트에서 언급해주신 `model`에 대한 분리를 이 영역에서부터 개선을 시작해보면 좋을거같아요.
우선 제가 [도메인(모델)](https://github.com/next-step/js-racingcar/pull/67#discussion_r841968863)을 분리하는 기준에 대해 자동차 미션을 기준으로 코멘트를 달아둔게 있어서 해당 코멘트를 읽어보시고 이어질 내용을 보시면 좋을거같아요.

다른 리뷰어님들이 남겨주신 코멘트에서도 분명 좋은 힌트가 있을거라고 생각하니 여유가 되신다면 다른 PR의 코멘트도 읽어보시는걸 추천드릴게요.

분리에 대한 생각은 위 PR로 갈음하고 현재 미션의 스텝1 에서는 라우팅을 담당하는 `navigation`(`currentTab`) 모델과 상품인 `product`, 해당 상품을 관리하는 `product manager`으로 분리해볼 수 있을거같아요.

cyrpess 테스트 파일을 분리해주신 것 처럼 모델을 똑같이 분리해보면 좋을거같아요.
브라우저에 의존하지 않고도 jest등의 테스트 툴을 이용해서 단위 테스트가 가능하도록 구성해보고 여유가 되신다면 테스트까지 작성해보면 좋을거같습니다!

리팩터링을 시작하기 어려우시다면 코멘트 후 DM주시면 빠르게 답변드리겠습니다 🙂

EungyuCho commented on cypress/integration/productManage.spec.js:
```diff
@@ -0,0 +1,142 @@
+describe('상품 추가하기 탭 관련 테스트', () => {
+  beforeEach(() => {
+    cy.visit('/');
+    cy.get(`#${SELECTOR.PRODUCT_MANAGE_MENU_ID}`).click();
```
첫 화면이 상품 관리라서 호출안해도 될거같아요

EungyuCho replied:
그리고 페이지를 렌더링 하는 부분은 이렇게 cy.get(navId).click()으로 호출하는 것 보다 [커맨드](https://docs.cypress.io/api/cypress-api/custom-commands#Usage)로 빼두면 더 직관적으로 테스트를 할 수 있을거같아요.

EungyuCho commented on cypress/integration/productManage.spec.js:
```diff
@@ -0,0 +1,142 @@
+          );
+        });
+    });
+  });
```
```suggestion

    it('상품 가격은 100이상 정수여야 합니다.', () => {
      const alertStub = cy.stub();
      cy.on('window:alert', alertStub);

      cy.get(`#${SELECTOR.PRODUCT_NAME_INPUT_ID}`).type(product.name);
      cy.get(`#${SELECTOR.PRODUCT_PRICE_INPUT_ID}`).type(product.price - 10);
      cy.get(`#${SELECTOR.PRODUCT_QUANTITY_INPUT_ID}`).type(product.quantity);
      cy.get(`#${SELECTOR.PRODUCT_ADD_BUTTON_ID}`)
        .click()
        .then(() => {
          expect(alertStub.getCall(0)).to.be.calledWith(
            ERROR_MESSAGE.PRODUCT_PRICE_HAVE_TO_OVER_100,
          );
        });
    });

    it('상품 가격은 10으로 나누어 떨어져야 합니다.', () => {
      const alertStub = cy.stub();
      cy.on('window:alert', alertStub);

      cy.get(`#${SELECTOR.PRODUCT_NAME_INPUT_ID}`).type(product.name);
      cy.get(`#${SELECTOR.PRODUCT_PRICE_INPUT_ID}`).type(product.price + 1);
      cy.get(`#${SELECTOR.PRODUCT_QUANTITY_INPUT_ID}`).type(product.quantity);
      cy.get(`#${SELECTOR.PRODUCT_ADD_BUTTON_ID}`)
        .click()
        .then(() => {
          expect(alertStub.getCall(0)).to.be.calledWith(
            ERROR_MESSAGE.PRODUCT_PRICE_HAVE_TO_DIVIDED_BY_10,
          );
        });
    });
  });
```
테스트에 활용하는 데이터를 선언해두고 사용하면 보다 직관적일거같아요.

EungyuCho commented on cypress.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "video": false,
+  "baseUrl": "http://localhost:8080",
+  "cypress-watch-and-reload": {
+    "watch": ["src/*", "src/controllers/*", "src/models/*", "src/utils/*", "src/views/*"]
+  }
```
이런 설정도 있었네요.. 👍

EungyuCho commented on src/VendingMachine.js:
```diff
@@ -0,0 +1,83 @@
+    this.vendingMachineView = new VendingMachineView(target);
+    this.validator = new Validator();
+
+    this.state = {
```
전반적으로 field가 Private로 관리될 부분이 많을거같아요.
아래 코멘트를 보시고 모델을 분리하신다면 private를 왜 써야하는지 생각해보시고 현수님이 클래스의 필드로 적절하다고 생각되는 접근지정자를 사용해주시면 되겠습니다~

EungyuCho reviewed (changes requested):
안녕하세요 현수님. 자판기 미션을 함께하게 된 조은규라고 합니다~

index 파일을 이용해서 path를 정리해주시려는점이나 전반적으로 코드 퀄리티가 좋다고 느꼈어요.  👍 👍 
하지만 본문 코멘트에 언급해주셨듯이 모델을 조금 더 분리해보면 좋겠다고 느꼈고 관련해서 코멘트에 제 생각을 남겨드렸어요 🙂

이번 미션에서 혼란스러운 부분에 대해 답을 찾고 가실 수 있으면 좋겠어요. 애매한 부분이 있으시다면 최대한 도와드릴게요! 미션 작성하시느라 고생하셨고 피드백 확인 부탁드립니다~

junghyeonsu committed:
chore: eslint 규칙 수정

- mjs 확장자 사용 가능하도록 변경

junghyeonsu committed:
chore: cypress.json 수정

- watch 옵션 수정

junghyeonsu committed:
feat: Product 도메인 생성

- name, price, quantity 프로퍼티를 가지고 있음
- 인스턴스 생성시에 validateProductDatas를 실행해서 검증

junghyeonsu committed:
feat: ProductManager 도메인 생성

- Product를 관리하는 ProductManager 생성

junghyeonsu committed:
feat: Router 도메인 생성

- route를 관리하는 도메인 생성

junghyeonsu committed:
chore: domains 폴더 import 관리

junghyeonsu committed:
chore: template들 변경

- export default가 아닌 export로 내보냄 (이름을 명시해주는 것이 좋다고 판단)
- 화살표 함수로 작성 (논리적인 순서로 함수 라인을 정하기 위해서)

junghyeonsu committed:
chore: store getValue 메서드 수정

- defaultValue를 넣어서 로컬스토리지에서 받아오지 못했으면 defaultValue를 뱉도록 수정

junghyeonsu committed:
chore: Validator 삭제

junghyeonsu committed:
refactor: VendingMachine 코드 리팩토링

- 각 도메인에 대한 처리들은 이동
- product를 객체로 보내줌
- route 처리에 대한 것도 도메인에 위임
- state 상태를 받아서 렌더링하는 방식 없앰

junghyeonsu committed:
refactor: VendingMachineView 리팩토링

junghyeonsu committed:
test: 불필요한 click 이벤트 삭제

junghyeonsu committed:
chore: 기준 상수 생성

junghyeonsu committed:
refactor: 중복된 이름 기능 리팩토링

junghyeonsu committed:
feat: 상품 추가 완료시 input 초기화

junghyeonsu committed:
test: alert 커맨드 생성

junghyeonsu committed:
refactor: hash router 생성

junghyeonsu commented on src/Router.mjs:
```diff
@@ -0,0 +1,21 @@
+  }
+}
+
+export default Router;
```
Router를 만들고 싶은데 현재 구조에서 어떻게 잘 구성을 해야할 지 느낌이 잘 오지 않습니다...ㅜ
Router에서 hashchange 이벤트를 관리하는 것이 깔끔해보이는데, 그렇게 하려면 Router에서 처음 rendering을 해주어야 하는데 현재 VendingMachine에서 도메인들을 부르고, 데이터를 가져오고 있습니다.
또 Router에서 도메인 인스턴스 객체를 생성하자니 뭔가 아닌 것 같고, template들을 가져와서 렌더링해주자니 VendingMachineView 인스턴스가 위치가 애매해지는 것 같고,,,

어떻게 해야할 지 감이 잘 잡히지 않는데... 혹시 조언을 주실 수 있을까요?

EungyuCho replied:
Router가 view를 생성자로 받으면 해결할 수 있지않을까요? 👀 
(아래 코멘트도 참고해주세요!)

junghyeonsu commented on src/VendingMachine.js:
```diff
@@ -0,0 +1,67 @@
+      );
+  }
+
+  initEvents() {
+    window.addEventListener('hashchange', this.handleHashChange.bind(this));
+    $(`#${SELECTOR.APP_ID}`).addEventListener('click', event => this.handleClickEvent(event));
+  }
+
+  handleClickEvent(event) {
+    const targetId = event.target.id;
+
+    if (targetId === SELECTOR.PRODUCT_ADD_BUTTON_ID) this.addProduct();
+  }
+
+  handleHashChange() {
+    const { hash: currentTab } = window.location;
+    this.vendingMachineView.renderTab(currentTab);
+    this.router.currentTab = currentTab;
+    this.dataBindingInTab();
+  }
+
+  addProduct() {
+    const product = {
+      name: $(`#${SELECTOR.PRODUCT_NAME_INPUT_ID}`).value,
+      price: $(`#${SELECTOR.PRODUCT_PRICE_INPUT_ID}`).value,
+      quantity: $(`#${SELECTOR.PRODUCT_QUANTITY_INPUT_ID}`).value,
+    };
+
+    try {
+      this.productManager.addProduct(product);
+    } catch (error) {
+      alert(error.message);
+      return;
+    }
+
+    this.vendingMachineView.resetProductForm();
+    this.vendingMachineView.renderProductTableInProductManageTab(this.productManager.products);
+  }
+}
```
현재 `VendingMachine`에서 도메인들을 관리하고 있고, View도 관리하고 있는 컨트롤러 같은 느낌이라
나중에가면 엄청 커질 것 같습니다.

근데 잘 나눠졌는지는 잘 모르겠습니다.. 😢  왜냐하면 모든 도메인들이 `VendingMachine`에 있다보니까 렌더링을 할 때도 도메인 데이터가 필요하고, 추가하는 로직도 도메인 데이터가 필요한데
분리를 하자니 분리한 곳에서 도메인 인스턴스를 또 선언해서 사용하려니 두번 선언이 되어서 나누기가 참 애매한 것 같습니다..
지금 약간 갈피를 잘 못잡고 있는 것 같습니다.., 멘토분이 보시기엔 어떤가요?

EungyuCho replied:
말씀해서 중복해서 선언하면 확실히 도메인을 분리하는 효과가 많이 떨어질거같아요.

애매한 부분에 동의하기에 개선이 필요하다고 생각되네요 🤔 
각 도메인에 어떤 객체들의 의존성이 필요한지 한번 정리해보시는거도 좋을거같아요.

제가 생각하기에 현재 두가지 방법이 있을거같은데요.
1. 생성자를 통해 각 모델에 필요한 의존성을 넣어준다.
2. [Custom Event Dispatch](https://ko.javascript.info/dispatch-events)를 통해 최상위 모델 데이터를 가지고 있는 최상위 VendingMachine.js에서 렌더링 데이터를 넣어준다.

저는 개인적으로 2번을 추천드리고 싶어요 😅 

etc) 구조를 보다가 번외로 생각해본 부분인데 모델을 먼저 생성한 뒤 view에 생성자로 모델의 데이터 넣어보면 어떨까요?

junghyeonsu commented on src/VendingMachineView.js:
```diff
@@ -0,0 +1,56 @@
+    $('main').innerHTML = RenderTab[currentTab]();
+  }
+
+  renderTabWithData(curretTab, data) {
+    const view = {
+      [HASH.PRODUCT_MANAGE_TAB]: () => {
+        this.renderProductTableInProductManageTab(data);
+      },
+      [HASH.VENDING_MACHINE_MANAGE_TAB]: () => {},
+      [HASH.PRODUCT_PURCHASE_TAB]: () => {},
+    };
+
+    view[curretTab](data);
+  }
```
위의 `renderTab` 메서드 이후에 해당 탭에 필요한 데이터들을 주입해주기 위해서 만든 메서드입니다.
이것 또한 이게 최선인지 잘 모르겠습니다..

EungyuCho replied:
저는 지금 상태에서 괜찮은 것 같은데 혹시 어떤 부분에 대해 확신이 없으신걸까요?? 👀 
확장성을 충분히 생각하시고 작성해주셨다고 생각해요

junghyeonsu commented on src/domains/Product.mjs:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  get name() {
+    return this.#name;
+  }
+
+  getProduct() {
+    return {
+      name: this.#name,
+      price: this.#price,
+      quantity: this.#quantity,
+    };
+  }
+
+  #validateProductDatas({ name, price, quantity }) {
+    this.#validateProductName(name);
+    this.#validateProductPrice(price);
+    this.#validateProductQuantity(quantity);
+  }
+
+  #validateProductName(productName) {
+    if (!productName) throw new Error(ERROR_MESSAGE.REQUIRED_PRODUCT_NAME);
+  }
+
+  #validateProductPrice(productPrice) {
+    if (!productPrice) throw new Error(ERROR_MESSAGE.REQUIRED_PRODUCT_PRICE);
+    if (productPrice < STANDARD.PRODUCT_PRICE_MINIMUM)
+      throw new Error(ERROR_MESSAGE.PRODUCT_PRICE_HAVE_TO_OVER_100);
+    if (productPrice % STANDARD.PRODUCT_PRICE_DIVIDE_BY !== 0)
+      throw new Error(ERROR_MESSAGE.PRODUCT_PRICE_HAVE_TO_DIVIDED_BY_10);
+  }
+
+  #validateProductQuantity(productQuantity) {
+    if (!productQuantity) throw new Error(ERROR_MESSAGE.REQUIRED_PRODUCT_QUANTITY);
+    if (productQuantity < STANDARD.PRODUCT_QUANTITY_MINIMUM)
+      throw new Error(ERROR_MESSAGE.PRODUCT_QUANTITY_HAVE_TO_OVER_1);
+  }
+}
+
+export default Product;
```
`Product` 도메인입니다. Product를 생성할 때 검증을 해주어서 오류를 뱉도록 했습니다.

junghyeonsu commented on src/domains/ProductManager.mjs:
```diff
@@ -0,0 +1,31 @@
+import { STORE_KEY } from '../constants.js';
+import Product from './Product.mjs';
+
+import store from '../utils/store.js';
+
+class ProductManager {
+  #products;
+
+  constructor() {
+    this.#products = store
+      .getValue(STORE_KEY.PRODUCTS, [])
+      .map(({ name, price, quantity }) => new Product({ name, price, quantity }));
+  }
+
+  get products() {
+    return this.#products.map(product => product.getProduct());
+  }
+
+  addProduct({ name, price, quantity }) {
+    const newProduct = new Product({ name, price, quantity });
+    this.#fliterSameNameProduct(name);
+    this.#products = [...this.#products, newProduct];
+    store.setValue(STORE_KEY.PRODUCTS, this.products);
+  }
+
+  #fliterSameNameProduct(productName) {
+    this.#products = this.#products.filter(product => product.name !== productName);
+  }
+}
+
+export default ProductManager;
```
`Product` 인스턴스를 관리하는 `ProductManager` 인스턴스입니다.

junghyeonsu committed:
chore: TAB 상수 삭제

junghyeonsu commented on src/domains/Router.mjs:
```diff
@@ -0,0 +1,21 @@
+import store from '../utils/store.js';
+import { STORE_KEY, HASH } from '../constants.js';
+
+class Router {
+  #currentTab;
+
+  constructor() {
+    this.#currentTab = store.getValue(STORE_KEY.CURRENT_TAB, HASH.PRODUCT_MANAGE_TAB);
+  }
+
+  get currentTab() {
+    return this.#currentTab;
+  }
+
+  set currentTab(tab) {
+    this.#currentTab = tab;
+    store.setValue(STORE_KEY.CURRENT_TAB, tab);
+  }
+}
+
+export default Router;
```
hash 값을 이용해서 라우터를 구현했는데, 해당 도메인은 hash를 받아서 currentTab이 어디인지 확인해주는 목적으로 만든 도메인입니다. 근데 hash 라우터를 잘 이용했는지 처음이라 잘 모르겠습니다.

멘토님은 어떠신가요?.. 적절하다고 생각하시나요?

EungyuCho replied:
currentTab을 관리하는 용도를 제외하고는 라우터 클래스가 하고있는게 없다보니 애매하네요 💦 
위에서 말씀해주신대로 hash change 이벤트를 해당 클래스에서 관리해줘야할거같아요.

junghyeonsu commented on src/utils/store.js:
```diff
@@ -0,0 +1,11 @@
+  setValue(key, value) {
+    localStorage.setItem(key, JSON.stringify(value));
+  },
+  getValue(value, defaultValue) {
+    const localStorageValue = localStorage.getItem(value);
+    return JSON.parse(localStorageValue) || defaultValue;
+  },
```
default value를 인자로 받아서 로컬스토리지에 존재하지 않으면 defaultValue를 return하도록 했습니다.

junghyeonsu reviewed (commented):
리뷰주신 내용들을 최대한 반영해봤는데
최대한 도메인을 나누려고 했고, 해시 라우터를 적용을 했는데
잘 적용이 됐는지 모르겠습니다... 우선 의견나눠보고 싶어서 리뷰 요청드립니다!
아래에 제가 바꾼 부분들에 대해서 코멘트를 남겨놨습니다!
참고해주세요!

EungyuCho reviewed (changes requested):
안녕하세요 현수님~

리팩터링 후에 각 도메인별로 잘 분리해주신거같아요.
충분히 잘 작성해주셨고 구조에 대한 부분은 다시 야이가 나눠보면서 개선을 한번 해보고 넘어가면 좋을거 같아요 🙂 

DM 확인 부탁드려요~

junghyeonsu committed:
chore: prettier printWidth 변경

- 100에서 120으로 변경

junghyeonsu committed:
chore: renderTabWithData 변경

- view 객체 전부 한줄로 표현

junghyeonsu committed:
chore: controllers 폴더 생성

junghyeonsu committed:
chore: VendingMachine 모듈 이름 변경

- VendingMachine에서 VendingMachineController로 변경

junghyeonsu committed:
refactor: MVC 패턴 구축

- domains -> models 변경
- 컨트롤러 세분화
- 뷰 인스턴스에서 단순 객체로 변경

EungyuCho reviewed (approved):
안녕하세요 현수님~

우선 다음 스텝을 진행해보면서 조금 더 구조에 대해 고민해보는게 좋을 것 같아서 머지하도록 하겠습니다~
고생하셨어요! 🙂
